scripts/ASTImportTest.sh

```diff
@@ -1,105 +1,165 @@
+
+set -euo pipefail
 
-set -e
-
-# Bash script to test the ast-import option of the compiler by
-# first exporting a .sol file to JSON, then loading it into the compiler
-# and exporting it again. The second JSON should be identical to the first
 READLINK=readlink
 if [[ "$OSTYPE" == "darwin"* ]]; then
     READLINK=greadlink
 fi
 REPO_ROOT=$(${READLINK} -f "$(dirname "$0")"/..)
 SOLIDITY_BUILD_DIR=${SOLIDITY_BUILD_DIR:-${REPO_ROOT}/build}
-SOLC=${SOLIDITY_BUILD_DIR}/solc/solc
-SPLITSOURCES=${REPO_ROOT}/scripts/splitSources.py
+SOLC=$SOLIDITY_BUILD_DIR/solc/solc
+SPLITSOURCES=$REPO_ROOT/scripts/splitSources.py
+
+# shellcheck source=scripts/common.sh
+source "${REPO_ROOT}/scripts/common.sh"
+
+function print_usage
+{
+  fail "Usage: ${0} ast [--exit-on-error]."
+}
+
+IMPORT_TEST_TYPE=
+ABORT_ON_ERROR=0
+for PARAM in "$@"
+do
+  case "$PARAM" in
+    ast) IMPORT_TEST_TYPE="ast" ;;
+    --exit-on-error) ABORT_ON_ERROR=1 ;;
+    *) print_usage ;;
+  esac
+done
 
 SYNTAXTESTS_DIR="${REPO_ROOT}/test/libsolidity/syntaxTests"
 ASTJSONTESTS_DIR="${REPO_ROOT}/test/libsolidity/ASTJSON"
-NSOURCES="$(find "$SYNTAXTESTS_DIR" -type f | wc -l)"
-
-# DEV_DIR="${REPO_ROOT}/../tmp/contracts/"
-# NSOURCES="$(find $DEV_DIR -type f | wc -l)" #TODO use find command
 
 FAILED=0
 UNCOMPILABLE=0
 TESTED=0
 
-if [[ "$(find . -maxdepth 0 -type d -empty)" == "" ]]; then
-    echo "Test directory not empty. Skipping!"
-    exit 1
-fi
+function ast_import_export_equivalence
+{
+    local sol_file="$1"
+    local input_files=( "${@:2}" )
 
-# function tests whether exporting and importing again leaves the JSON ast unchanged
+    # save ast json as expected result (silently)
+    if ! $SOLC --combined-json ast --pretty-json --json-indent 4 "${input_files[@]}" > expected.json 2> stderr_export.txt
+    then
+        printError "ERROR: AST export failed for input file $sol_file."
+        printError ""
+        printError "Compiler stderr:"
+        cat ./stderr_export.txt >&2
+        printError ""
+        printError "Compiler stdout:"
+        cat ./expected.json >&2
+        return 1
+    fi
+
+    # import it, and export it again as obtained result (silently)
+    if ! $SOLC --import-ast --combined-json ast --pretty-json --json-indent 4 expected.json > obtained.json 2> stderr_import.txt
+    then
+        printError "ERROR: AST import failed for input file $sol_file"
+        printError ""
+        printError "Compiler stderr:"
+        cat ./stderr_import.txt >&2
+        printError ""
+        printError "Compiler stdout:"
+        cat ./obtained.json >&2
+        return 1
+    fi
+
+#    if [[ $((TESTED % 50)) == 1 ]]
+#    then
+#      echo "!" >> expected.json
+#    fi
+    if diff_files expected.json obtained.json
+    then
+        echo -n "✅"
+    else
+        printError "❌ ERROR: AST reimport failed for ${sol_file}"
+        if [[ $ABORT_ON_ERROR == 1 ]]
+        then
+            printError "You can find the files used for this test here: $(pwd)"
```

Comment:
I think something like that is useful. It allows to investigate the resulting json files and gives information what was actually done in these tests. so this should allow an easy reproduction.